pkg/entity/tournament.go

```diff
@@ -0,0 +1,855 @@
+package entity
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"gorm.io/datatypes"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	realtime "github.com/domino14/liwords/rpc/api/proto/realtime"
+)
+
+type TournamentGame struct {
+	Scores        []int                           `json:"s"`
+	Results       []realtime.TournamentGameResult `json:"r"`
+	GameEndReason realtime.GameEndReason          `json:"g"`
+}
+
+const Unpaired = -1
+
+type Pairing struct {
+	Players  []string                        `json:"p"`
+	Games    []*TournamentGame               `json:"g"`
+	Outcomes []realtime.TournamentGameResult `json:"o"`
+}
+
+type PlayerRoundInfo struct {
+	Pairing *Pairing `json:"p"`
+	Record  []int    `json:"r"`
+	Spread  int      `json:"s"`
+}
+
+type Standing struct {
+	Player string
+	Wins   int
+	Losses int
+	Draws  int
+	Spread int
+}
+
+type TournamentManager interface {
+	GetPlayerRoundInfo(string, int) (*PlayerRoundInfo, error)
+	SubmitResult(int, string, string, int, int, realtime.TournamentGameResult,
+		realtime.TournamentGameResult, realtime.GameEndReason, bool, int) error
+	GetStandings(int) ([]*Standing, error)
+	SetPairing(string, string, int) error
+	StartRound(int) error
+	IsRoundComplete(int) (bool, error)
+	IsFinished() (bool, error)
+	Serialize() (datatypes.JSON, error)
+}
+
+type PairingMethod int
+
+const (
+	Random PairingMethod = iota
+	RoundRobin
+	KingOfTheHill
+	Elimination
+	// Need to implement eventually
+	// Swiss
+	// Performance
+
+	// Manual simply does not make any
+	// pairings at all. The director
+	// has to make all the pairings themselves.
+	Manual
+)
+
+type TournamentClassic struct {
+	Matrix         [][]*PlayerRoundInfo `json:"t"`
+	Players        []string             `json:"p"`
+	PlayerIndexMap map[string]int       `json:"i"`
+	PairingMethods []PairingMethod      `json:"m"`
+	GamesPerRound  int                  `json:"g"`
+}
+
+type TournamentType int
+
+const (
+	ClassicTournamentType TournamentType = iota
+	// It's gonna be lit:
+	ArenaTournamentType
+)
+
+type TournamentPersons struct {
+	Persons map[string]int `json:"p"`
+}
+
+type TournamentControls struct {
+	GameRequest    *realtime.GameRequest `json:"g"`
+	PairingMethods []PairingMethod       `json:"p"`
+	NumberOfRounds int                   `json:"n"`
+	GamesPerRound  int                   `json:"g"`
+	Type           TournamentType        `json:"t"`
+	StartTime      time.Time             `json:"s"`
+}
+
+type Tournament struct {
+	sync.RWMutex
+	UUID              string              `json:"u"`
+	Name              string              `json:"n"`
+	Description       string              `json:"e"`
+	Directors         *TournamentPersons  `json:"d"`
+	Players           *TournamentPersons  `json:"p"`
+	Controls          *TournamentControls `json:"c"`
+	IsStarted         bool                `json:"i"`
+	TournamentManager TournamentManager   `json:"m"`
```

Comment:
we need an array of these for multiple divisions

Reply:
done.